DESIGN.md

```diff
@@ -0,0 +1,756 @@
+improvement. Consider a read API that takes `T: AsMut<[u8]>` for the buffer
+argument.
+
+```rust
+async fn read<T: AsMut<[u8]>>(&self, dst: T) { ... }
+
+struct MyUnsableBuf {
+    mem: [u8; 10],
+}
+
+impl AsMut<[u8]> {
+    fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.mem[..]
+    }
+}
+```
+
+This read function takes ownership of the buffer; however, any pointer to the
+buffer obtained from a value becomes invalid when the value moves. Storing the
+buffer value at a stable location while the operation is in-flight should be
+sufficient to satisfy safety.
+
+## Closing resources
+
+Idiomatically with Rust, closing a resource is performed in the drop handler,
+and within an asynchronous context, the drop handler should be non-blocking.
+Closing an io-uring resource requires canceling any in-flight operations, which
+is an asynchronous process. Consider an open TcpStream associated with
+file-descriptor (FD) 10. A task submits a read operation to the kernel, but the
+`TcpStream` is dropped and closed before it sees it. A new `TcpStream` is
+accepted, and the kernel reuses FD 10. At this point, the kernel sees the
+original read operation request with FD 10 and completes it on the new
+`TcpStream`, not the intended `TcpStream`. The runtime receives the completion
+of the read operation. It discards the result because the associated operation
+future is gone, resulting in the caller losing data from the second
+TcpStream. This problem occurs even issuing a cancellation request for the read
+operation. There is no guarantee the kernel will see the cancellation request
+before completing the operation.
+
+There are two options for respecting both requirements, neither ideal: closing
+the resource in the background or blocking the thread in the drop handler. If
+the resource is closed in the background, the process may encounter unexpected
+errors, such as "too many open files." Blocking the thread to cancel in-flight
+operations and close the resource prevents the runtime from processing other
+tasks and adds latency across the system.
+
+Instead, tokio-uring will provide an explicit asynchronous close function on
+resource types.
+
+```rust
+impl TcpStream {
+    async fn close(self) { ... }
+}
+
+my_tcp_stream.close().await;
+```
+
+The resource must still tolerate the caller dropping it without being explicitly
+closed. In this case, tokio-uring will close the resource in the background,
+avoiding blocking the runtime. The drop handler will move ownership of the
+resource handle to the runtime and submit cancellation requests for any
+in-flight operation. Once all existing in-flight operations complete, the
+runtime will submit a close operation.
+
+If the drop handler must process closing a resource in the background, it will
+notify the developer by emitting a warning message using [`tracing`]. In the
+future, it may be possible for Rust to provide a `#[must_not_drop]` attribute.
```

Comment:
You are correct re: panicking, which means drop must always be safe and correct, though not necessarily ideal. The idea of `#[must_not_drop]` is that it would be a lint and not enforced.

Just like with `#[must_use]` lints, if the developer means to drop and let cleanup happen in the background, they can silence the error (via some TBD mechanism).